File: 3.x.x/crates-utopia/messaging/src/adapter/chat/discord.rs
//! PHP `Utopia\Messaging\Adapter\Chat\Discord`.

use serde_json::json;
use url::Url;

use crate::adapter::{expect_discord, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::Discord as DiscordMessage;
use crate::php::php_empty_str;
use crate::response::{Response, ResponseData};

/// PHP `Adapter\Chat\Discord`.
#[derive(Debug)]
pub struct Discord {
    base: AdapterBase,
    webhook_url: String,
    webhook_id: String,
}

impl Discord {
    /// PHP `__construct($webhookURL)`.
    pub fn new(webhook_url: impl Into<String>) -> Result<Self, MessagingError> {
        let webhook_url = webhook_url.into();
        let parsed = Url::parse(&webhook_url)
            .map_err(|_| MessagingError::invalid_argument("Invalid Discord webhook URL format."))?;
        if parsed.scheme() != "https" {
            return Err(MessagingError::invalid_argument(
                "Discord webhook URL must use HTTPS scheme.",
            ));
        }
        if parsed.host_str() != Some("discord.com") {
            return Err(MessagingError::invalid_argument(
                "Discord webhook URL must use discord.com as host.",
            ));
        }
        let path = parsed.path();
        let webhook_id = path
            .split("/webhooks/")
            .nth(1)
            .and_then(|rest| rest.split('/').next())
            .unwrap_or("")
            .to_string();
        if php_empty_str(&webhook_id) {
            return Err(MessagingError::invalid_argument(
                "Discord webhook ID cannot be empty.",
            ));
        }
        Ok(Self {
            base: AdapterBase::default(),
            webhook_url,
            webhook_id,
        })
    }

    fn process_discord(&self, message: &DiscordMessage) -> ResponseData {
        let mut query = Vec::new();
        if let Some(wait) = message.get_wait() {
            query.push(("wait", if wait { "1".into() } else { "0".into() }));
        }
        if let Some(thread_id) = message.get_thread_id() {
            query.push(("thread_id", thread_id.to_string()));
        }
        let qs = if query.is_empty() {
            String::new()
        } else {
            let mut ser = url::form_urlencoded::Serializer::new(String::new());
            for (k, v) in &query {
                ser.append_pair(k, v);
            }
            format!("?{}", ser.finish())
        };

        let result = self.request_default(
            "POST",
            &format!("{}{qs}", self.webhook_url),
            &["Content-Type: application/json".into()],
            Some(json!({
                "content": message.get_content(),
                "username": message.get_username(),
                "avatar_url": message.get_avatar_url(),
                "tts": message.get_tts(),
                "embeds": message.get_embeds(),
                "allowed_mentions": message.get_allowed_mentions(),
                "components": message.get_components(),
                "attachments": message.get_attachments(),
                "flags": message.get_flags(),
                "thread_name": message.get_thread_name(),
            })),
        );

        let mut response = Response::new("chat");
        if (200..300).contains(&result.status_code) {
            response.set_delivered_to(1);
            response.add_result(&self.webhook_id, "");
        } else if (400..500).contains(&result.status_code) {
            response.add_result(&self.webhook_id, "Bad Request.");
        } else {
            response.add_result(&self.webhook_id, "Unknown Error.");
        }
        response.to_array()
    }
}

impl Adapter for Discord {
    fn get_name(&self) -> &'static str {
        "Discord"
    }
    fn get_type(&self) -> &'static str {
        "chat"
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::Discord
    }
    fn get_max_messages_per_request(&self) -> usize {
        1
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(
            self.process_discord(expect_discord(message)?),
        ))
    }
}

File: packages/messaging/src/Adapter/Chat/Discord.php
<?php

namespace Utopia\Messaging\Adapter\Chat;

use InvalidArgumentException;
use Utopia\Messaging\Adapter;
use Utopia\Messaging\Messages\Discord as DiscordMessage;
use Utopia\Messaging\Response;

class Discord extends Adapter
{
    protected const NAME = 'Discord';
    protected const TYPE = 'chat';
    protected const MESSAGE_TYPE = DiscordMessage::class;
    protected string $webhookId = '';

    /**
     * @param  string  $webhookURL Your Discord webhook URL.
     * @throws InvalidArgumentException When webhook URL is invalid
     */
    public function __construct(
        private readonly string $webhookURL,
    ) {
        parent::__construct();
        // Validate URL format
        if (!filter_var($webhookURL, FILTER_VALIDATE_URL)) {
            throw new InvalidArgumentException('Invalid Discord webhook URL format.');
        }

        // Validate URL uses https scheme
        $urlParts = parse_url($webhookURL);
        if (!isset($urlParts['scheme']) || $urlParts['scheme'] !== 'https') {
            throw new InvalidArgumentException('Discord webhook URL must use HTTPS scheme.');
        }

        // Validate host is discord.com
        if (!isset($urlParts['host']) || $urlParts['host'] !== 'discord.com') {
            throw new InvalidArgumentException('Discord webhook URL must use discord.com as host.');
        }

        // Extract and validate webhook ID
        $parts = explode('/webhooks/', $urlParts['path']);
        if (\count($parts) >= 2) {
            $webhookParts = explode('/', $parts[1]);
            $this->webhookId = $webhookParts[0];
        }
        if ($this->webhookId === '' || $this->webhookId === '0') {
            throw new InvalidArgumentException('Discord webhook ID cannot be empty.');
        }
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getType(): string
    {
        return static::TYPE;
    }

    public function getMessageType(): string
    {
        return static::MESSAGE_TYPE;
    }

    public function getMaxMessagesPerRequest(): int
    {
        return 1;
    }

    /**
     * @return array{deliveredTo: int, type: string, results: array<array<string, mixed>>}
     *
     * @throws \Exception
     */
    protected function process(DiscordMessage $message): array
    {
        $query = [];

        if (!\is_null($message->getWait())) {
            $query['wait'] = $message->getWait();
        }
        if (!\is_null($message->getThreadId())) {
            $query['thread_id'] = $message->getThreadId();
        }

        $queryString = $query === [] ? '' : '?' . http_build_query($query);

        $response = new Response($this->getType());
        $result = $this->request(
            method: 'POST',
            url: "{$this->webhookURL}{$queryString}",
            headers: [
                'Content-Type: application/json',
            ],
            body: [
                'content' => $message->getContent(),
                'username' => $message->getUsername(),
                'avatar_url' => $message->getAvatarUrl(),
                'tts' => $message->getTTS(),
                'embeds' => $message->getEmbeds(),
                'allowed_mentions' => $message->getAllowedMentions(),
                'components' => $message->getComponents(),
                'attachments' => $message->getAttachments(),
                'flags' => $message->getFlags(),
                'thread_name' => $message->getThreadName(),
            ],
        );

        $statusCode = $result['statusCode'];

        if ($statusCode >= 200 && $statusCode < 300) {
            $response->setDeliveredTo(1);
            $response->addResult($this->webhookId);
        } elseif ($statusCode >= 400 && $statusCode < 500) {
            $response->addResult($this->webhookId, 'Bad Request.');
        } else {
            $response->addResult($this->webhookId, 'Unknown Error.');
        }

        return $response->toArray();
    }
}
